crates/sui-json-rpc/src/read_api.rs

```diff
@@ -74,10 +75,14 @@ impl RpcReadApiServer for ReadApi {
             .collect())
     }
 
-    async fn get_object(&self, object_id: ObjectID) -> RpcResult<GetObjectDataResponse> {
+    async fn get_object(
```

Comment:
I think it might make a lot more sense to leave this API as is, and add a new API called `get_archived_object_version` or something similar. This would:

- Keep this API simpler - it can continue returning the exists/not-exists/deleted enum.
- Simplify the implementation of `get_object_read` which is now very confusing.
- Make it clearer to the user that they are trying to access an archive, not live state, which may be pruned.
- Simplify the error return values.

However, If you choose to continue with the current approach, may I then suggest the following:
- Merge `SequenceNumberTooHigh` into `NotExists`.
- Rename `ExistsButPastNotFound` to `ObjectVersionPruned`

Reply:
@mystenmark  Split into two functions. `ExistsButPastNotFound` renamed to `VersionNotFound`. I kept `VersionTooHigh` as it's trivial to tell and may be helpful to callers